SpotifyAPI.Docs/docs/SpotifyWebAPI/auth.md

```diff
@@ -100,6 +102,277 @@ static void auth_OnResponseReceivedEvent(Token token, string state, string error
+    AutoRefresh = true
+};
+webApiFactory.OnAuthSuccess += (sender, e) => authorized = true;
+
+try
+{
+    spotify = await webApiFactory.GetWebApiAsync();
+}
+catch (Exception ex)
+{
+    UpdateStatus($"Spotify failed to load: {ex.Message}");
+}
+```
+
+#### Way 2 - Synchronous
+This will halt the program until your SpotifyWebAPI is available.
+
+```c#
+SecureWebAPIFactory webApiFactory;
+SpotifyWebAPI spotify;
+
+bool requestComplete = false;
+
+Task.Factory.StartNew(async () =>
+{
+    // You should store a reference to WebAPIFactory if you are using AutoRefresh or want to manually
+    // refresh it later on. New WebAPIFactory objects cannot refresh a SpotifyWebAPI that they did not
+    // give to you.
+    webApiFactory = new SecureWebAPIFactory("INSERT LINK TO YOUR index.php HERE")
+    {
+        Scope = Scope.UserReadPrivate | Scope.UserReadEmail | Scope.PlaylistReadPrivate | Scope.UserLibraryRead | Scope.UserReadPrivate | Scope.UserFollowRead | Scope.UserReadBirthdate | Scope.UserTopRead | Scope.PlaylistReadCollaborative | Scope.UserReadRecentlyPlayed | Scope.UserReadPlaybackState | Scope.UserModifyPlaybackState | Scope.PlaylistModifyPublic,
+        AutoRefresh = true
+    };
+    webApiFactory.OnAuthSuccess += (sender, e) => authorized = true;
+
+    try
+    {
+        spotify = await webApiFactory.GetWebApiAsync();
+    }
+    catch (Exception ex)
+    {
+        UpdateStatus($"Spotify failed to load: {ex.Message}");
+    }
+
+    requestComplete = true;
+});
+
+while (!requestComplete) ;
+```
+
+### Verbose way - using the SecureAuthorizationCodeAuth class
+Since the SecureWebAPIFactory not only simplifies the whole process but offers additional functionality too
+(such as AutoRefresh and AuthSuccess AuthFailure events), use of this way is very verbose and is only
+recommended if you are having issues with SecureWebAPIFactory or need access to the tokens.
+
+```c#
+SecureAuthorizationCodeAuth auth = new SecureAuthorizationCodeAuth(
+    exchangeServerUri: "INSERT LINK TO YOUR index.php HERE",
+    serverUri: "http://localhost:4002",
+    scope: Scope.UserReadPrivate | Scope.UserReadEmail | Scope.PlaylistReadPrivate | Scope.UserLibraryRead | Scope.UserReadPrivate | Scope.UserFollowRead | Scope.UserReadBirthdate | Scope.UserTopRead | Scope.PlaylistReadCollaborative | Scope.UserReadRecentlyPlayed | Scope.UserReadPlaybackState | Scope.UserModifyPlaybackState | Scope.PlaylistModifyPublic
+);
+auth.AuthReceived += async (sender, response) =>
+{
+    lastToken = await auth.ExchangeCodeAsync(response.Code);
+
+    spotify = new SpotifyWebAPI()
+    {
+        TokenType = lastToken.TokenType,
+        AccessToken = lastToken.AccessToken
+    };
+
+    authenticated = true;
+    auth.Stop();
+};
+auth.OnAccessTokenExpired += async (sender, e) => spotify.AccessToken = (await auth.RefreshAuthAsync(lastToken.RefreshToken)).AccessToken;
+auth.Start();
+auth.OpenBrowser();
+```
+
+## Creating the Exchange Server
```

Comment:
I would like to abstract this a little bit more in the docs and keep the PHP Script in a repo of your choice. (Also this should be ### AFAIK)

Sth like the following:
```
### Token Swap Endpoint

To use this method, an external HTTP Server needs to be able to supply the following HTTP Endpoints to our application:
`/swap` - Swaps out an `authorization_code` with an `access_token` and `refresh_token` - The following parameters are required in the JSON POST Body: ...
`/refresh` - Refreshes an `access_token` - The following parameters are required in the JSON POST Body: ...

The following open-source token swap endpoints can be used:
rollersteaam/spotify-token-swap-php
simontaen/SpotifyTokenSwap (note: would be cool if we could support 2 servers at once!)
```
I also don't think we need the redirect functionality on the server, every client should be able to build this URL themselves

Reply:
I have refactored the Creating the Exchange Server doc entry for your Token Swap Endpoint entry in this commit https://github.com/rollersteaam/SpotifyAPI-NET/commit/321724830f604d8fcb9275aa6136ff3818c4f864.

But I'm not sure if putting the redirect functionality on the server should be avoided or not, so until that's decided the PR won't be able to move forward yet, as if it shouldn't be avoided, some more information will need to be added to the docs (which I'll do as necessary).

Reply:
It should be fine in this state, will check again tho

Reply:
Regarding this quote:

> However the only reason why I have a redirect for the GET auth code request to Spotify in my php example is that it keeps the client_id and redirect_uri off a program so that anyone decompiling the program can't find that information.

Well, the information is in the login URL of spotify anyway, so instead of decompiling, a user just has to run the program. Also, `client_id` and `redirect_uri` can be public. However, the current `ImplicitGrantAuth` solution requires that you use localhost as the redirect URI, which allows everyone to use this "domain" and create `ImplicitGrantAuth` as well. This is fixed by your auth solution.

(The only drawback of this solution is, that every user can simply request an unlimited amount of tokens by hitting the endpoint often. However, this is normal and can't be avoided when having an anonymous client side app. So in the end, this solution is the best we can currently support 👍 )